go/stellar/bundle/boxer_test.go

```diff
@@ -13,314 +15,483 @@ import (
 		replaceWith = "b"
 	}
-	res.VisB64 = res.VisB64[:85] + replaceWith + res.VisB64[86:]
+	boxed.VisParentB64 = boxed.VisParentB64[:85] + replaceWith + boxed.VisParentB64[86:]
+
+	_, _, _, _, err = DecodeAndUnbox(m, ring, boxed.toBundleEncodedB64())
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "visible hash mismatch")
+}
+
+func TestBoxAndEncodeCatchesMalformedBundles(t *testing.T) {
+	bundle := testBundle(t)
+	ring := newPukRing()
+	pukSeed, pukGen := ring.makeGen(t, 3)
+
+	// put a different account and secret in the AccountBundle
+	newAcctID, newSecret, err := randomStellarKeypair()
+	require.NoError(t, err)
+	newAB := map[stellar1.AccountID]stellar1.AccountBundle{
+		newAcctID: stellar1.AccountBundle{
+			AccountID: newAcctID,
+			Signers:   []stellar1.SecretKey{newSecret},
+		},
+	}
+	bundle.AccountBundles = newAB
+
+	// encode should error because the bundle is invalid
+	_, err = BoxAndEncode(&bundle, pukGen, pukSeed)
+	require.Contains(t, err.Error(), "account in AccountBundles not in Accounts")
+}
 
-	dec2, err := Decode(res.EncB64)
+type accountCan struct {
+	accountID stellar1.AccountID
+	encB64    string
+}
+type canned struct {
+	pukSeedB64   string
+	pukGen       int
+	encParentB64 string
+	visParentB64 string
+	accounts     []accountCan
+}
+
+func (c *canned) puk(t *testing.T) (puk libkb.PerUserKeySeed) {
+	bs, err := base64.StdEncoding.DecodeString(c.pukSeedB64)
 	require.NoError(t, err)
+	require.Equal(t, len(puk), len(bs))
+	copy(puk[:], bs)
+	return puk
+}
 
-	_, _, err = Unbox(nil, dec2, res.VisB64, puk)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "hash mismatch")
+func (c *canned) gen() keybase1.PerUserKeyGeneration {
+	return keybase1.PerUserKeyGeneration(c.pukGen)
+}
+
+func (c *canned) toBundleEncodedB64() BundleEncoded {
+	benc := BundleEncoded{
+		EncParent:   c.encParentB64,
+		VisParent:   c.visParentB64,
+		AcctBundles: make(map[stellar1.AccountID]string),
+	}
+	for _, acct := range c.accounts {
+		benc.AcctBundles[acct.accountID] = acct.encB64
+	}
+	return benc
+}
+
+func (c *canned) ring(t *testing.T) *pukRing {
+	pukSeed := c.puk(t)
+	pukGen := c.gen()
+	return &pukRing{
+		map[keybase1.PerUserKeyGeneration]libkb.PerUserKeySeed{
+			pukGen: pukSeed,
+		},
+	}
+}
+
+var validAccounts = []accountCan{
+	{"GAWZ7HVPKRGCH2KP6475XV6HA2CAF44MXWWE5RKV4LMMGB6FNNSEPNPE", "hKFlxKUc3fqMOAYv9m6ycpnpQA4CSriSQoPVNbvsXXEgv4WsixkaTThkKGMlYuWvTJdHeqPdYXo/Mw156xq8MaIbzDTriFplFNcLhNYdi8f1ViHqvVIecX2frU/BOtIsAlqknnhl4+Z1u2kUnZYI5pZRvUV5H1loSWC4tBmEoCgK1S6XrLx1POOIiKkH8EFMXVrB6+BjbieW1w8HTXNp0jWbKkq+QoKz8MCjZ2VuA6FuxBibtfMD40wYqnfPsGePn1RphpzcEqv7mZehdgE="},
+	{"GBPLLHOKPRFSCE6FP6D37AUUNYCJ2TSPOHMHAR54TBDLWWPIZCD4RL3G", "hKFlxKX3zqweGMWyl4vOC8ht0jngDnTEpWqGBePh7okF9S003QStjI9Td1d+urqj/k4etwgLiPO8LHaaQ7o2WAwuSXakHJVJ2xIq+T+MYnoobbx5sArk4wsg7AWTX+uw/rXoyN7P9ZAgDSH+4oZ6/0j8dwR6RAoc9c+dog8xnW3eTkeSlj2KYx6XnEOOlOcCBCKsZePi9eoN92CxvB5MRN6tR7w3PH4yynqjZ2VuA6FuxBha4JmmVyysST3avwMxayHfTExp7tnLwHOhdgE="},
+}
```

Comment:
was this meant to be used in `cans` below for the `accounts` field?

Reply:
ahh. yes. you are correct. will remove.